File: src/main/java/com/sefa/utils/ShareDataUtils.java
package com.sefa.utils;

import com.opencsv.CSVReader;
import com.opencsv.exceptions.CsvException;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Share data utilities for stock price and currency rate processing
 * Converted from Python utils/share_data_utils.py
 */
public class ShareDataUtils {
    
    public static class TimedFmv {
        private final long entryTimeInMillis;
        private final double fmv;
        
        public TimedFmv(long entryTimeInMillis, double fmv) {
            this.entryTimeInMillis = entryTimeInMillis;
            this.fmv = fmv;
        }
        
        public long getEntryTimeInMillis() {
            return entryTimeInMillis;
        }
        
        public double getFmv() {
            return fmv;
        }
    }
    
    public static class TimedFmvWithInrRate extends TimedFmv {
        private final double inrRate;
        
        public TimedFmvWithInrRate(long entryTimeInMillis, double fmv, double inrRate) {
            super(entryTimeInMillis, fmv);
            this.inrRate = inrRate;
        }
        
        public double getInrRate() {
            return inrRate;
        }
    }
    
    private static final Map<String, List<TimedFmv>> priceMapCache = new HashMap<>();
    private static List<TimedFmv> usdInrCache = new ArrayList<>();
    
    /**
     * Validate dates for FMV lookup
     */
    private static void validateDates(long historicEntryTimeInMs, long desiredPurchaseTimeInMs, long usedFmvTimeInMs) {
        if (historicEntryTimeInMs > desiredPurchaseTimeInMs) {
            throw new IllegalArgumentException(
                String.format("Historical FMV date %s can NOT be newer than purchase date = %s",
                    DateUtils.logTimestamp(historicEntryTimeInMs),
                    DateUtils.logTimestamp(desiredPurchaseTimeInMs))
            );
        }
        
        long daysDiff = (DateUtils.lastWorkDayInMs(desiredPurchaseTimeInMs) - historicEntryTimeInMs) / (24 * 60 * 60 * 1000);
        
        if (daysDiff > 0) {
            String msg = String.format(
                "Historical FMV at %s was NOT available(maybe due to Public Holiday or weekends) last available data is %d days old(on %s)",
                DateUtils.logTimestamp(desiredPurchaseTimeInMs),
                (int) daysDiff,
                DateUtils.displayTime(historicEntryTimeInMs)
            );
            Logger.log(msg);
            Logger.log("Hence using the next available FMV at %s", DateUtils.logTimestamp(usedFmvTimeInMs));
        }
    }
    
    /**
     * Initialize USD/INR exchange rate cache
     */
    private static List<TimedFmv> initUsdInrMap() {
        if (usdInrCache.isEmpty()) {
            Logger.log("Parsing USD/INR exchange rate map");
            
            String scriptPath = System.getProperty("user.dir");
            Path usdInrPath = Paths.get(scriptPath, "historic_data", "usd_inr_price_history.csv");
            
            if (!FileUtils.fileExists(usdInrPath.toString())) {
                throw new IllegalArgumentException("USD/INR historical data NOT present at " + usdInrPath);
            }
            
            try (CSVReader reader = new CSVReader(new FileReader(usdInrPath.toFile()))) {
                List<String[]> records = reader.readAll();
                
                // Skip header row
                for (int i = 1; i < records.size(); i++) {
                    String[] row = records.get(i);
                    long entryTimeInMs = DateUtils.parseDateFromUsdInrFormat(row[0].replace("\"", ""));
                    String priceStr = row[1].replace("\"", "").replace(",", "");
                    double price = Double.parseDouble(priceStr);
                    
                    usdInrCache.add(new TimedFmv(entryTimeInMs, price));
                }
                
                // Sort by date for efficient lookup
                usdInrCache.sort(Comparator.comparingLong(TimedFmv::getEntryTimeInMillis));
                
            } catch (IOException | CsvException e) {
                throw new RuntimeException("Failed to read USD/INR data: " + e.getMessage(), e);
            }
        }
        
        return usdInrCache;
    }
    
    /**
     * Initialize price map for a ticker
     */
    private static List<TimedFmv> initMap(String ticker) {
        if (!priceMapCache.containsKey(ticker)) {
            Logger.log("Parsing FMV price map for ticker = %s", ticker);
            
            List<TimedFmv> tickerPriceMap = new ArrayList<>();
            String scriptPath = System.getProperty("user.dir");
            Path historicSharePath;
            
            // Use the new adobe_price_history.csv file for ADBE
            if ("adbe".equalsIgnoreCase(ticker)) {
                historicSharePath = Paths.get(scriptPath, "historic_data", "adobe_price_history.csv");
            } else {
                // Fallback to old structure for other tickers
                historicSharePath = Paths.get(scriptPath, "historic_data", "shares", ticker.toLowerCase(), "data.csv");
            }
            
            if (!FileUtils.fileExists(historicSharePath.toString())) {
                throw new IllegalArgumentException(
                    String.format("Historic share data for share %s NOT present at %s", ticker, historicSharePath)
                );
            }
            
            try (CSVReader reader = new CSVReader(new FileReader(historicSharePath.toFile()))) {
                List<String[]> records = reader.readAll();
                
                // Skip header row
                for (int i = 1; i < records.size(); i++) {
                    String[] row = records.get(i);
                    long entryTimeInMs;
                    double price;
                    
                    if ("adbe".equalsIgnoreCase(ticker)) {
                        // Handle adobe_price_history.csv format
                        entryTimeInMs = DateUtils.parseDateFromAdobeFormat(row[0]);
                        // Clean the Close/Last value (remove $ sign if present)
                        String closePriceStr = row[1].replace("$", "").replace(",", "");
                        price = Double.parseDouble(closePriceStr);
                    } else {
                        // Handle old data.csv format
                        entryTimeInMs = DateUtils.parseYyyyMmDd(row[0]).getTimeInMillis();
                        price = Double.parseDouble(row[4]); // Close column
                    }
                    
                    tickerPriceMap.add(new TimedFmv(entryTimeInMs, price));
                }
                
                // Sort by date for efficient lookup
                tickerPriceMap.sort(Comparator.comparingLong(TimedFmv::getEntryTimeInMillis));
                priceMapCache.put(ticker, tickerPriceMap);
                
            } catch (IOException | CsvException e) {
                throw new RuntimeException("Failed to read share data for " + ticker + ": " + e.getMessage(), e);
            }
        }
        
        return priceMapCache.get(ticker);
    }
    
    /**
     * Get Fair Market Value for a ticker at a specific time
     */
    public static double getFmv(String ticker, long purchaseTimeInMs) {
        Logger.debugLog("%s: Querying FMV at %s", ticker, DateUtils.displayTime(purchaseTimeInMs));
        
        List<TimedFmv> priceMap = initMap(ticker);
        TimedFmv previousEntry = null;
        
        for (TimedFmv entry : priceMap) {
            long entryTimeInMs = entry.getEntryTimeInMillis();
            if (entryTimeInMs >= purchaseTimeInMs) {
                if (entryTimeInMs > purchaseTimeInMs && previousEntry != null) {
                    validateDates(previousEntry.getEntryTimeInMillis(), purchaseTimeInMs, entryTimeInMs);
                    return entry.getFmv();
                }
                return entry.getFmv();
            }
            previousEntry = entry;
        }
        
        // Updated error message to reflect new file locations
        String tickerSharePrice = "adbe".equalsIgnoreCase(ticker) 
            ? "historic_data/adobe_price_history.csv"
            : String.join("/", "historic_data", "shares", ticker, "data.csv");
        
        throw new IllegalArgumentException(
            String.format("No FMV data for share ticker %s in %s for date %s",
                ticker, tickerSharePrice, DateUtils.logTimestamp(purchaseTimeInMs))
        );
    }
    
    /**
     * Get USD to INR exchange rate for a given timestamp
     */
    public static double getUsdInrRate(long timeInMs) {
        List<TimedFmv> usdInrData = initUsdInrMap();
        TimedFmv previousEntry = null;
        
        for (TimedFmv entry : usdInrData) {
            long entryTimeInMs = entry.getEntryTimeInMillis();
            if (entryTimeInMs >= timeInMs) {
                if (entryTimeInMs > timeInMs) {
                    return previousEntry != null ? previousEntry.getFmv() : entry.getFmv();
                }
                return entry.getFmv();
            }
            previousEntry = entry;
        }
        
        // If no future data found, use the last available rate
        if (previousEntry != null) {
            return previousEntry.getFmv();
        }
        
        throw new IllegalArgumentException(
            "No USD/INR rate data available for date " + DateUtils.logTimestamp(timeInMs)
        );
    }
    
    /**
     * Get closing price for a ticker at end time
     */
    public static double getClosingPrice(String ticker, long endTimeInMs) {
        List<TimedFmv> priceMap = initMap(ticker);
        
        Optional<TimedFmv> result = priceMap.stream()
            .filter(price -> price.getEntryTimeInMillis() <= endTimeInMs)
            .max(Comparator.comparingLong(TimedFmv::getEntryTimeInMillis));
        
        return result.map(TimedFmv::getFmv)
            .orElseThrow(() -> new IllegalArgumentException("No closing price data found for " + ticker));
    }
    
    /**
     * Get peak price in INR within a time range
     */
    public static double getPeakPriceInInr(String ticker, long startTimeInMs, long endTimeInMs) {
        if (startTimeInMs > endTimeInMs) {
            throw new IllegalArgumentException(
                String.format("start_time_in_ms = %d is greater than equal to end_time_in_ms = %d",
                    startTimeInMs, endTimeInMs)
            );
        }
        
        List<TimedFmv> priceMap = initMap(ticker);
        String currencyCode = TickerMapping.getTickerCurrencyInfo(ticker);
        
        // Filter prices within the time range
        List<TimedFmv> filteredPrices = priceMap.stream()
            .filter(price -> price.getEntryTimeInMillis() >= startTimeInMs && 
                           price.getEntryTimeInMillis() <= endTimeInMs)
            .collect(Collectors.toList());
        
        // Convert to prices with INR rates
        List<TimedFmvWithInrRate> priceMapWithInrRate = filteredPrices.stream()
            .map(price -> {
                double inrRate = "USD".equals(currencyCode) 
                    ? getUsdInrRate(price.getEntryTimeInMillis())
                    : 1.0; // Fallback for other currencies
                return new TimedFmvWithInrRate(price.getEntryTimeInMillis(), price.getFmv(), inrRate);
            })
            .collect(Collectors.toList());
        
        // Find maximum value
        TimedFmvWithInrRate maxValue = priceMapWithInrRate.stream()
            .max(Comparator.comparingDouble(price -> price.getFmv() * price.getInrRate()))
            .orElseThrow(() -> new IllegalArgumentException("No price data found in the given range"));
        
        double peakPriceInInr = maxValue.getFmv() * maxValue.getInrRate();
        
        Logger.debugLogJson(Map.of(
            "start_time", DateUtils.displayTime(startTimeInMs),
            "end_time", DateUtils.displayTime(endTimeInMs),
            "max_fmv($)", maxValue.getFmv(),
            "max_fmv($)_at", DateUtils.displayTime(maxValue.getEntryTimeInMillis()),
            "inr_conversion_rate", maxValue.getInrRate(),
            "effective_price(INR)", peakPriceInInr
        ));
        
        Logger.log("Peak price for ticker = %s from %s to %s is %.2f INR at rate %.2f INR/USD",
            ticker, DateUtils.displayTime(startTimeInMs), DateUtils.displayTime(endTimeInMs),
            peakPriceInInr, maxValue.getInrRate());
        
        return peakPriceInInr;
    }
} 

File: utils/share_data_utils.py
from utils.runtime_utils import warn_missing_module

warn_missing_module("pandas")
import pandas as pd
import os
import typing as t

from . import date_utils, logger
from .ticker_mapping import ticker_currency_info
from .rates import rbi_rates_utils


def __validate_dates(
    historic_entry_time_in_ms: int,
    desired_purchase_time_in_ms: int,
    used_fmv_time_in_ms: int,
):
    if historic_entry_time_in_ms > desired_purchase_time_in_ms:
        raise AssertionError(
            f"Historical FMV date {date_utils.log_timestamp(historic_entry_time_in_ms)} "
            + "can NOT be newer than purchase date "
            + f"= {date_utils.log_timestamp(desired_purchase_time_in_ms)}"
        )
    days_diff = (
        date_utils.last_work_day_in_ms(desired_purchase_time_in_ms)
        - historic_entry_time_in_ms
    ) / (24 * 60 * 60 * 1000)

    date_utils.last_work_day_in_ms(desired_purchase_time_in_ms)

    if days_diff > 0:
        msg = (
            f"Historical FMV at {date_utils.log_timestamp(desired_purchase_time_in_ms)} "
            + "was NOT available(maybe due to Public Holiday or weekends) last available data is "
            + f"{int(days_diff)} days old(on {date_utils.display_time(historic_entry_time_in_ms)})"
        )
        logger.log(msg)
        logger.log(
            f"Hence using the next available FMV at {date_utils.log_timestamp(used_fmv_time_in_ms)}"
        )
        # if days_diff > 2:
        #     raise Exception(msg)


TimedFmv = t.TypedDict("TimedFmv", {"entry_time_in_millis": int, "fmv": float})


TimedFmvWithInrRate = t.TypedDict(
    "TimedFmvWithInrRate",
    {
        "entry_time_in_millis": int,
        "fmv": float,
        "inr_rate": float,
    },
)


price_map_cache: t.Dict[str, t.List[TimedFmv]] = {}
usd_inr_cache: t.List[TimedFmv] = []


def __parse_date_from_adobe_format(date_str: str) -> int:
    """
    Parse date from adobe_price_history.csv format: MM/DD/YYYY
    """
    from datetime import datetime
    date_time = datetime.strptime(date_str, "%m/%d/%Y")
    return date_utils.epoch_in_ms(date_time)


def __parse_date_from_usd_inr_format(date_str: str) -> int:
    """
    Parse date from usd_inr_price_history.csv format: DD-MM-YYYY
    """
    from datetime import datetime
    date_time = datetime.strptime(date_str, "%d-%m-%Y")
    return date_utils.epoch_in_ms(date_time)


def __init_usd_inr_map() -> t.List[TimedFmv]:
    """
    Initialize USD to INR exchange rate cache
    """
    global usd_inr_cache
    if not usd_inr_cache:
        print("Parsing USD/INR exchange rate map")
        script_path = os.path.realpath(os.path.dirname(__file__))
        usd_inr_path = os.path.join(
            script_path,
            os.pardir,
            "historic_data",
            "usd_inr_price_history.csv",
        )
        if not os.path.exists(usd_inr_path):
            raise AssertionError(
                f"USD/INR historical data NOT present at {usd_inr_path}"
            )
        df = pd.read_csv(usd_inr_path)

        for _, data in df.iterrows():
            entry_time_in_ms = __parse_date_from_usd_inr_format(data["Date"])
            # Clean the price value (remove any formatting)
            price_str = str(data["Price"]).replace(',', '')
            usd_inr_cache.append(
                {"entry_time_in_millis": entry_time_in_ms, "fmv": float(price_str)}
            )

        # Sort by date for efficient lookup
        usd_inr_cache.sort(key=lambda x: x["entry_time_in_millis"])

    return usd_inr_cache


def __init_map(ticker: str) -> t.List[TimedFmv]:
    if ticker not in price_map_cache:
        print(f"Parsing FMV price map for ticker = {ticker}")
        ticker_price_map: t.List[TimedFmv] = []
        script_path = os.path.realpath(os.path.dirname(__file__))
        
        # Use the new adobe_price_history.csv file
        if ticker.lower() == "adbe":
            historic_share_path = os.path.join(
                script_path,
                os.pardir,
                "historic_data",
                "adobe_price_history.csv",
            )
        else:
            # Fallback to old structure for other tickers
        historic_share_path = os.path.join(
            script_path,
            os.pardir,
            "historic_data",
            "shares",
            ticker.lower(),
            "data.csv",
        )
        
        if not os.path.exists(historic_share_path):
            raise AssertionError(
                f"Historic share data for share {ticker} NOT present at {historic_share_path}"
            )
        df = pd.read_csv(historic_share_path)

        for _, data in df.iterrows():
            if ticker.lower() == "adbe":
                # Handle adobe_price_history.csv format
                entry_time_in_ms = __parse_date_from_adobe_format(data["Date"])
                # Clean the Close/Last value (remove $ sign if present)
                close_price_str = str(data["Close/Last"]).replace('$', '').replace(',', '')
                price = float(close_price_str)
            else:
                # Handle old data.csv format
            entry_time_in_ms = date_utils.parse_yyyy_mm_dd(data["Date"])[
                "time_in_millis"
            ]
                price = data["Close"]
            
            ticker_price_map.append(
                {"entry_time_in_millis": entry_time_in_ms, "fmv": price}
            )

        # Sort by date for efficient lookup
        ticker_price_map.sort(key=lambda x: x["entry_time_in_millis"])
        price_map_cache[ticker] = ticker_price_map

    return price_map_cache[ticker]


def get_fmv(ticker: str, purchase_time_in_ms: int) -> float:
    logger.debug_log(
        f"{ticker}: Querying FMV at {date_utils.display_time(purchase_time_in_ms)}"
    )

    previous_entry_data = None
    for entry_data in __init_map(ticker):
        entry_time_in_ms = entry_data["entry_time_in_millis"]
        if entry_time_in_ms >= purchase_time_in_ms:
            if entry_time_in_ms > purchase_time_in_ms:
                previous_entry_time_in_ms = previous_entry_data["entry_time_in_millis"]
                __validate_dates(
                    previous_entry_time_in_ms, purchase_time_in_ms, entry_time_in_ms
                )
                return entry_data["fmv"]
            return entry_data["fmv"]

        previous_entry_data = entry_data
    
    # Updated error message to reflect new file locations
    if ticker.lower() == "adbe":
        ticker_share_price = "historic_data/adobe_price_history.csv"
    else:
    ticker_share_price = os.path.join("historic_data", "shares", ticker, "data.csv")
    
    raise AssertionError(
        f"No FMV data for share ticker {ticker} in {ticker_share_price} for date "
        + f"{date_utils.log_timestamp(purchase_time_in_ms)}"
    )


def get_usd_inr_rate(time_in_ms: int) -> float:
    """
    Get USD to INR exchange rate for a given timestamp
    """
    usd_inr_data = __init_usd_inr_map()
    
    previous_entry_data = None
    for entry_data in usd_inr_data:
        entry_time_in_ms = entry_data["entry_time_in_millis"]
        if entry_time_in_ms >= time_in_ms:
            if entry_time_in_ms > time_in_ms:
                if previous_entry_data:
                    return previous_entry_data["fmv"]
                else:
                    return entry_data["fmv"]
            return entry_data["fmv"]
        previous_entry_data = entry_data
    
    # If no future data found, use the last available rate
    if previous_entry_data:
        return previous_entry_data["fmv"]
    
    raise AssertionError(
        f"No USD/INR rate data available for date "
        + f"{date_utils.log_timestamp(time_in_ms)}"
    )


def get_closing_price(ticker: str, end_time_in_ms: int) -> float:
    price_map = list(
        filter(
            lambda price: price["entry_time_in_millis"] <= end_time_in_ms,
            sorted(
                __init_map(ticker),
                key=lambda price: price["entry_time_in_millis"],
                reverse=True,
            ),
        )
    )

    return price_map[0]["fmv"]


def get_peak_price_in_inr(
    ticker: str, start_time_in_ms: int, end_time_in_ms: int
) -> float:
    if start_time_in_ms > end_time_in_ms:
        raise AssertionError(
            f"start_time_in_ms = {start_time_in_ms} is greater "
            + f"than equal to end_time_in_ms = {end_time_in_ms}"
        )

    price_map = list(
        filter(
            lambda price: price["entry_time_in_millis"] <= end_time_in_ms
            and price["entry_time_in_millis"] >= start_time_in_ms,
            sorted(
                __init_map(ticker),
                key=lambda price: price["entry_time_in_millis"],
                reverse=True,
            ),
        )
    )
    
    # Use the new USD/INR rate function or fallback to RBI rates
    price_map_with_inr_rate: t.Iterator[TimedFmvWithInrRate] = map(
        lambda price: {
            **price,
            "inr_rate": get_usd_inr_rate(price["entry_time_in_millis"])
            if ticker_currency_info[ticker] == "USD"
            else rbi_rates_utils.get_rate_for_prev_mon_for_time_in_ms(
                ticker_currency_info[ticker], price["entry_time_in_millis"]
            ),
        },
        price_map,
    )

    max_value = max(
        price_map_with_inr_rate, key=lambda price: price["fmv"] * price["inr_rate"]
    )

    peak_price_in_inr = max_value["fmv"] * max_value["inr_rate"]

    logger.debug_log_json(
        {
            "start_time": date_utils.display_time(start_time_in_ms),
            "end_time": date_utils.display_time(end_time_in_ms),
            "max_fmv($)": max_value["fmv"],
            "max_fmv($)_at": date_utils.display_time(max_value["entry_time_in_millis"]),
            "inr_conversion_rate": max_value["inr_rate"],
            "effective_price(INR)": peak_price_in_inr,
        }
    )

    logger.log(
        f"Peak price for ticker = {ticker} from {date_utils.display_time(start_time_in_ms)} "
        + f"to {date_utils.display_time(end_time_in_ms)} is {peak_price_in_inr} "
        + f"INR at rate {max_value['inr_rate']} INR/USD"
    )

    return max_value["fmv"] * max_value["inr_rate"]
